### DISCUSSION TITLE
Cannot Boot After Spoofing EFI Revision

### DISCUSSION THREAD

**Discussion by startergo** [dakanji/RefindPlus#25] [General]
**Describe the bug**
No booting occurs through latest few commits and no log file is generated

**To Reproduce**
Steps to reproduce the behaviour:
Try to boot any OS

**Expected behaviour**
Expected to see the RP bootpicker or the login screen

**Environment:**
 - RefindPlus Version : 0.13.0.AA (Pre-Release)
 - Hardware
   * Computer         :  MacPro 5,1]
   * GPU Model        :  RX 580, GT-640 Mac EFI
   * GPU Port         : DP
   * GPU Flash        : YES-MacVidCards and regular RX-580. With or without MVC card same result except with the MVC card gray screen remains.
 - Affected OS        : Any Mac OS

**Log and Config Files**
No Log file is generated.

**Additional context**
Restored working RP (sorry don't know which revision), but chain loading with the MVC GT-640 OC Hangs. Attached please find the OC log file hang:
[Archive.zip](https://github.com/dakanji/RefindPlus/files/6039453/Archive.zip)

After soft blessing OC directly it boots to the login screen. Attached the log file:
[opencore-2021-02-24-231046.txt.zip](https://github.com/dakanji/RefindPlus/files/6039456/opencore-2021-02-24-231046.txt.zip)

Sorry the working RP version is the RELEASE version so no RP log file is generated. Here it is attached. Maybe you can find out what version it is:
[Refind_x64.efi.zip](https://github.com/dakanji/RefindPlus/files/6039463/Refind_x64.efi.zip)

**Comment by dakanji** [comment]
> Sorry the working RP version is the RELEASE version

Just to be clear, do you mean RELEASE works but DEBUG (same otherwise) does not?

**Comment by startergo** [comment]
> > Sorry the working RP version is the RELEASE version
> 
> Just to be clear, do you mean RELEASE works but DEBUG (same otherwise) does not?

No, I just don't have a backup of DEBUG version to provide the log. This comment only reflects the OC hang during chainloading. RP not booting is for both RELEASE and DEBUG versions (latest few commits I don't remember where it started)

**Comment by dakanji** [comment]
Bug is said to be that RP is not booting and not logging since the last few commits:
> No booting occurs through latest few commits and no log file is generated

I have just tested again and it works for me.
Can you clarify what the exact issue is as a bit lost?

It is that you don't see the boot screen when running RP?
Or you do and it hangs when boot Mac OS?

**Comment by startergo** [comment]
it is not booting and no logging occurs. It is similar to what I had with the MBP with the memory leak except there is no logging at all. If I use the MVC card I stay on the grey screen forever. With a regular card there is a black screen forever. I don't get to the boot selector at all.

**Comment by dakanji** [comment]
I see.

Try the attached current build. 
If it also doesn't work, attach the last log file you have.

[X110-BOOTx64.efi.zip](https://github.com/dakanji/RefindPlus/files/6042593/X110-BOOTx64.efi.zip)

**Comment by startergo** [comment]
I just tried the rEFInd 0.13.01 and while it boots there were no partitions to select just the tools. Here is the config file which is just a copy of the RP file renamed
[refind.conf.zip](https://github.com/dakanji/RefindPlus/files/6042732/refind.conf.zip)

**Comment by dakanji** [comment]
OK. Please test the previously attached file.

**Comment by startergo** [comment]
Same thing with the attached build. I have no previous undeleted logs.

**Comment by dakanji** [comment]
OK. Thanks.

Might be worth opening a ticket upstream since it seems RP appears to be behaving consistently with rEFInd.

**Comment by dakanji** [comment]
> Same thing with the attached build. 

One odd thing though ... the attached should not be able get to showing tools without outputting a log of progress up to that point.

**Comment by startergo** [comment]
> > Same thing with the attached build.
> 
> One odd thing though ... the attached should not be able get to showing tools without outputting a log of progress up to that point.

I only see tools in refind nothing in RP.

**Comment by dakanji** [comment]
> I only see tools in refind nothing in RP.

OK. Either way though, there is an issue there.

Maybe test rEFInd with the RP specific tokens removed from the config to make sure there is no conflict. If there is still an issue, then best to open the ticket: https://sourceforge.net/p/refind/tickets/ (Seems tickets might be disabled. Might need to go into the forum)

**Comment by CornCat78** [comment]
@dakanji just tested X110 as requested.  RP initiated fine and loaded bootpicker on my cMP5,1.  See RP logs attached.
[21d25m1645.log](https://github.com/dakanji/RefindPlus/files/6042876/21d25m1645.log)

**Comment by dakanji** [comment]
Thanks @CornCat78 

@startergo Initially looks like it might be something specific to you.
rEFInd 0.13.1 has logging. Maybe activate this to gain some insight. If it runs up to the tools display, there should be output to show what's happening. I think Rod Smith will request this anyway.

**Comment by startergo** [comment]
> @dakanji just tested X110 as requested. RP initiated fine and loaded bootpicker on my cMP5,1. See RP logs attached.
> [21d25m1645.log](https://github.com/dakanji/RefindPlus/files/6042876/21d25m1645.log)

Can you please try the touch refind config file?just rename it config.conf.

**Comment by dakanji** [comment]
> Can you please try the touch refind config file?

Good idea.

@CornCat78 ... he means to try using the config file here: https://github.com/dakanji/RefindPlus/issues/24#issuecomment-785836680

Since you are using a MyBootMgr setup, best way will be to temporarily replace your overrides.conf file with this and reboot.

**Comment by CornCat78** [comment]
@dakanji - just to be clear, test with the replacement `overrides.conf ` on X110?  Either way, unfortunately it will have to be later as my cMP5,1 is in use for work at the moment.  Will try this and report back as soon as I have no more commitments for the day.

**Comment by CornCat78** [comment]
@dakanji 
@startergo 
Sorry for the late reply; it's been a long day at work. Anyway, using X110 build as well as rEFInd configuration file in place of overrides.conf, cMP5,1 boots and am presented with a rEFInd graphical boot picker.  Logs are attached.
[21d25x1607.log](https://github.com/dakanji/RefindPlus/files/6046063/21d25x1607.log)

**Comment by startergo** [comment]
This is the rEFInd log:

[refind.log.zip](https://github.com/dakanji/RefindPlus/files/6047147/refind.log.zip)

**Comment by startergo** [comment]
Dayo I found what is "wrong". I will let you find it too through the log. Let me know if you find it (hint: I changed something :) )
[refind.log.zip](https://github.com/dakanji/RefindPlus/files/6047208/refind.log.zip)

**Comment by dakanji** [comment]
Noted the EFI Revision change you made.
Remember to always eliminate your changes before concluding on issues.

**Comment by startergo** [comment]
I still consider this a bug even though it is in the upstream rEFInd. This was the only way to initialize an NVIDIA GOP without chain-loading OC through RP on systems with EFI revision 1.1. There is a change in the code between 0.12.0-0.13.1 of rEFInd, which breaks this workaround and leaves the system in unbootable state. @joevt what do you think?

**Comment by joevt** [comment]
@startergo So you enabled the hack that allows new GOPs that want EFI 2.3 to work. Somehow that hack causes a hang.
I don't think much, but I suppose it would be nice if that hack that worked before continued to work.

Is the hack being applied by RefindPlus or by my FakeUEFI2.efi and ReloadPCIRom.efi driver? I believe the RefindPlus implementations are slightly different but I didn't do much comparing or testing of either.

**Comment by startergo** [comment]
> @startergo So you enabled the hack that allows new GOPs that want EFI 2.3 to work. Somehow that hack causes a hang.
> I don't think much, but I suppose it would be nice if that hack that worked before continued to work.
> 
> Is the hack being applied by RefindPlus or by my FakeUEFI2.efi and ReloadPCIRom.efi driver? I believe the RefindPlus implementations are slightly different but I didn't do much comparing or testing of either.

This was addressed in the RP code earlier so it is not needed when chain-loading OC through RP, but if one uses OC alone without this driver early in the booting process NVIDIA's GOP does not initialize. It  broke in the rEFInd code between versions 0.12 and 0.13.1 and Dayo merged the upstream rEFInd code and now both bootloaders hang with this driver loaded.

**Comment by dakanji** [comment]
> I still consider this a bug

A broken hack is not a bug.

You basically have the following options:
- Implement the hack and never chain-load OC.
- Do not implement the hack and use RP to chainload OC.
- Amend the hack to account for the current reality. That is, add a kill switch to revert it when not booting into OC.
- ID the breaking change for consideration to restore the previous situation (if it doesn't cause other issues)
  * The main change between 0.12.0 and 0.13.0 was to eliminate a memory allocation issue. Can't compromise this for a hack
- Eliminate the need for the hack by implementing the feature in OC

**Comment by startergo** [comment]
> * Eliminate the need for the hack by implementing the feature in OC

This is the best approach.

**Comment by dakanji** [comment]
> This is the best approach.

Correct. Note that this can be done by loading the driver in OC instead of injecting it in firmware.

**Reply by dakanji** [reply]
"Probably be done"

**Reply by startergo** [reply]
I tested it before and it did not work as it needed to be loaded early in the booting process or reloaded on demand like in RP.

**Reply by dakanji** [reply]
There are challenges there indeed. Not straightforward to implement by embedding in their code. 

Have to consider some of the other options listed instead.

**Reply by dakanji** [reply]
Theoretically though, the actions can still be done in a driver.
It's just that the current driver needs to be expanded.

**Comment by startergo** [comment]
At present the best option is chain loading OC through RP. Days may I request for your consideration a self blessing feature? Like R
P  finding itself after NVRAM reset?

**Reply by dakanji** [reply]
> self blessing feature

Not sure how that would work or if even possible.
I don't think even OC can do this if NVRAM is reset outside OC.
Any thoughts @joevt?

**Reply by startergo** [reply]
OC is pretty consistent in finding itself after NVRAM reset now.

**Reply by dakanji** [reply]
That's not OC doing anything special. This is just how the UEFI specs work.

If you install RP as a fallback efi (BOOTx64 file in EFI/BOOT), as per MyBootMgr recommended setup for instance, it will be loaded after an NVRAM reset.

Specific "self blessing" is something else beyond this. Implies finding itself even when in a location and/or with a filename inconsistent with the specs.

**Reply by startergo** [reply]
> If you install RP as a fallback efi (BOOTx64 file in EFI/BOOT), as per MyBootMgr recommended setup for instance, it will be loaded after an NVRAM reset.

In multi disk configuration?

**Reply by dakanji** [reply]
> In multi disk configuration?

The first fallback efi found (OC, RP, rEFInd, Clover, whatever) will always be loaded.
We've touched on this before: https://forums.macrumors.com/posts/29413652

**Comment by dakanji** [comment]
Synced with an upstream change that impacts on EFI driver loading. 
Something to do with fixing a 0.13.1 secure boot bug.
Not sure if relevant for your case.  
Doubtful but who knows?

**Reply by startergo** [reply]
Very strange. The debug does not boot to RP bootpicker, the release boots but does not chainload OC. Otherwise it boots a regular Mac OS/Windows.
